crates/snp_validation/src/lib.rs

```diff
@@ -287,96 +308,132 @@ impl Snp {
+            bail!("snp policy has higher abi minor than firmware");
         }
 
         // Check reserved fields
         if report.body.rsvd1 != 0 || report.body.rsvd3 != 0 || report.body.rsvd4 != 0 {
-            return Err(anyhow!("snp report reserved fields were set"));
+            bail!("snp report reserved fields were set");
         }
 
         for value in report.body.rsvd2 {
-            if value != 0 {
-                return Err(anyhow!("snp report reserved fields were set"));
-            }
+            ensure!(value == 0, "snp report reserved fields were set");
         }
 
         for value in report.body.rsvd5 {
-            if value != 0 {
-                return Err(anyhow!("snp report reserved fields were set"));
-            }
+            ensure!(value == 0, "snp report reserved fields were set");
         }
 
         for value in report.body.policy.rsvd {
-            if value != 0 {
-                return Err(anyhow!("snp report policy reserved fields were set"));
-            }
+            ensure!(value == 0, "snp report policy reserved fields were set");
         }
 
         for value in report.body.plat_info.rsvd {
-            if value != 0 {
-                return Err(anyhow!("snp report platform_info reserved fields were set"));
-            }
+            ensure!(
+                value == 0,
+                "snp report platform_info reserved fields were set"
+            );
         }
 
         // Check fields not set by Enarx
-        for value in report.body.author_key_digest {
-            if value != 0 {
-                return Err(anyhow!(
-                    "snp report author_key_digest field not set by Enarx"
-                ));
-            }
+        for value in report.body.host_data {
+            ensure!(value == 0, "snp report host_data field not set by Enarx");
         }
 
-        for value in report.body.host_data {
-            if value != 0 {
-                return Err(anyhow!("snp report host_data field not set by Enarx"));
-            }
+        ensure!(report.body.vmpl == 0, "snp report vmpl field invalid value");
+
+        // Check field set by Enarx
+        for value in report.body.report_id_ma {
+            ensure!(
+                value == 255,
+                "snp report report_id_ma field not the value set by Enarx"
+            );
         }
 
-        for value in report.body.id_key_digest {
-            if value != 0 {
-                return Err(anyhow!("snp report id_key_digest field not set by Enarx"));
+        if let Some(config) = config {
+            ensure!(
+                config.abi.matches(&report.body.policy.into()),
+                "snp minimum abi not met"
+            );
+
+            if !config.measurements.signer.is_empty() {
+                let signed = config
+                    .measurements
+                    .signer
+                    .iter()
+                    .any(|s| report.body.author_key_digest == s.as_ref());
```

Comment:
If you used a `HashSet`, then this would be a very simple `config.measurements.signer.contains(report.body.author_key_digest)` call

Same in most of these

Reply:
Would be nice. I've tried it, and get stuck where the report fields (`[u8;32]`, or some other size) cannot be converted into `Digest`.

* I can't get `impl From<&[u8]> for Digest` to work properly, the compiler complains that `impl From<[u8;32]> for &Digest` is not implemented (nor can I implement that, it seems, without borrower protest).
* Also, converting the report field array/slice `to_vec()` inadvertently picks up `der::encode::Encode::to_vec()`, which is definitely bad (and unfortunate to have such a function name collision).

Reply:
Fixed that in the diff I sent you